lightning/src/routing/router.rs

```diff
@@ -1564,45 +1564,61 @@ fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters,
 	for path in route.paths.iter_mut() {
 		let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
 
-		// Choose the last publicly known node as the starting point for the random walk
-		if let Some(starting_hop) = path.iter().rev().find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.pubkey))) {
-			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.pubkey);
+		// Remember the last three nodes of the random walk and avoid looping back on them.
+		// Init with the last three nodes from the actual path, if possible.
+		let last_id = NodeId::from_pubkey(&path.last().unwrap().pubkey);
+		let second_last_id = NodeId::from_pubkey(&path.get(path.len().saturating_sub(2)).unwrap().pubkey);
+		let third_last_id = NodeId::from_pubkey(&path.get(path.len().saturating_sub(3)).unwrap().pubkey);
+		let mut nodes_to_avoid: [NodeId; 3] = [third_last_id, second_last_id, last_id];
+		let mut nodes_ring_index = 0;
+
+		// Choose the last publicly known node as the starting point for the random walk.
+		let mut cur_hop: Option<NodeId> = None;
+		let mut path_nonce = [0u8; 12];
+		if let Some(starting_hop) = path.iter().rev()
+			.find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.pubkey))) {
+				cur_hop = Some(NodeId::from_pubkey(&starting_hop.pubkey));
+				path_nonce.copy_from_slice(&cur_hop.unwrap().as_slice()[..12]);
+		}
+
+		// Init PRNG with the path-dependant nonce, which is static for private paths.
+		let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
+		let mut random_path_bytes = [0u8; ::core::mem::size_of::<usize>()];
 
-			// Init PRNG with path nonce
-			let mut path_nonce = [0u8; 12];
-			path_nonce.copy_from_slice(&cur_node_id.as_slice()[..12]);
-			let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
-			let mut random_path_bytes = [0u8; ::core::mem::size_of::<usize>()];
+		// Pick a random path length in [1 .. 3]
+		prng.process_in_place(&mut random_path_bytes);
+		let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
```

Comment:
Think there was a similar comment in another PR, but no need for the wrapping variants here given the use of `usize` and `3`.

Reply:
Yes, we had this discussion before, and I agree that 'normal' arithmetic can be more readable in many circumstances. However, in instances like this where already some function is applied I personally prefer the 'functional style' readability allowing to read the chained operations left to right over mixing the two like `(usize::from_be_bytes(random_path_bytes) % 3) + 1`.

Reply:
Ah, makes sense. Could also use the non-wrapping versions `rem` and `add`. No need to change though as I think we want to just merge now.